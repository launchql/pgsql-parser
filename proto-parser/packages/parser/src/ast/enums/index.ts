export * from './enums';
export * from './enums-json';