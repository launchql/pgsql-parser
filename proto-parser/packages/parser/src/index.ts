export * from './store';
export * from './parser';
export * from './options';
export * from './utils';