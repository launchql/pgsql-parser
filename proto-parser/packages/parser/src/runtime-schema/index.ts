export { RuntimeSchemaGenerator } from './generator';
export type { NodeSpec, FieldSpec, RuntimeSchemaOptions } from './types';
