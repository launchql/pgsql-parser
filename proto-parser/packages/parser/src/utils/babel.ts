import generate from '@babel/generator';
import * as t from '@babel/types';

export const convertAstToCode = (body: any[]) => {
  const ast = t.file(t.program(body));
  const { code } = generate(ast);
  return code;
};

export const createDefaultImport = (importName: string, source: string) => {
  return t.importDeclaration(
    [t.importDefaultSpecifier(t.identifier(importName))],
    t.stringLiteral(source)
  );
};

export const createNamedImport = (importNames: string[], source: string) => {
  const specifiers = importNames.map(name =>
    t.importSpecifier(t.identifier(name), t.identifier(name))
  );
  return t.importDeclaration(specifiers, t.stringLiteral(source));
};

export const createNamedImportAsSuffix = (importNames: string[], source: string, suffix: string) => {
  const specifiers = importNames.map(name =>
    t.importSpecifier(t.identifier(name + suffix), t.identifier(name))
  );
  return t.importDeclaration(specifiers, t.stringLiteral(source));
};
