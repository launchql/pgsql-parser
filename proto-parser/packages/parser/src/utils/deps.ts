import { Type, Enum, Field } from '@launchql/protobufjs';
import { isPrimitiveType } from './types';
import { NODE_TYPE } from '../constants';

const dependenciesResolve = (types: Type[], enums: Enum[], symbolName: string, resolved: string[], unresolved: string[]) => {
    if (symbolName === NODE_TYPE) return;

    unresolved.push(symbolName);

    let dep = [...types, ...enums].find(dep => dep.name === symbolName);
    if (!dep) {
        throw new Error(`Dependency Not Found ${symbolName}`)
    }

    const depAsType = types.find(t => t.name === dep.name);

    if (depAsType) {
        const fields = Object.values(depAsType.fields);
        for (let i = 0; i < fields.length; i++) {
            const childDep = fields[i];
            if (!resolved.includes(childDep.type)) {
                if (unresolved.includes(childDep.type)) {
                    // console.warn(`Circular reference detected ${symbolName}, ${dep}`);
                    continue;
                }
                if (!isPrimitiveType(childDep.type)) {
                    dependenciesResolve(types, enums, childDep.type, resolved, unresolved);
                }
            }
        }   
    }
    resolved.push(symbolName);
    const index = unresolved.indexOf(symbolName);
    unresolved.splice(index);
}

export const getDependencies = (name: string, types: Type[], enums: Enum[]) => {
    const resolved = [];
    const unresolved = [];
    dependenciesResolve(types, enums, name, resolved, unresolved);
    return resolved;
}

const dependentsResolve = (types, enums, symbolName, resolved, unresolved) => {
    if (symbolName === NODE_TYPE) return;

    unresolved.push(symbolName);

    [...types, ...enums].forEach(dep => {
        if (resolved.includes(dep.name) || unresolved.includes(dep.name)) {
            return;
        }

        const depAsType = types.find(t => t.name === dep.name);

        let isDependent = false;
        if (depAsType) {
            const fields: Field[] = Object.values(depAsType.fields);
            isDependent = fields.some(field => field.type === symbolName);
        } 

        if (isDependent) {
            if (unresolved.includes(dep.name)) {
                throw new Error(`Circular reference detected: ${symbolName} <-> ${dep.name}`);
            }

            dependentsResolve(types, enums, dep.name, resolved, unresolved);
        }
    });

    if (!resolved.includes(symbolName)) {
        resolved.push(symbolName);
    }

    const index = unresolved.indexOf(symbolName);
    if (index !== -1) {
        unresolved.splice(index, 1);
    }
};

export const getDependents = (symbolName, types, enums) => {
    const resolved = [];
    const unresolved = [];
    dependentsResolve(types, enums, symbolName, resolved, unresolved);
    return resolved.filter(name => name !== symbolName); // Exclude the symbol itself from its dependents
};
