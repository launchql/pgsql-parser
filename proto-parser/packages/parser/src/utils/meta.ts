import { toSpecialCamelCase } from './index';
import * as t from '@babel/types';

/**
 * Converts an AST (Abstract Syntax Tree) representation of a SQL query into
 * a meta-level AST that represents the code to generate the original AST.
 * This meta-level AST can then be used to programmatically construct the AST
 * for similar SQL queries. Essentially, it's an AST that describes how to
 * build an AST, enabling dynamic code generation based on the structure of
 * the original SQL AST. 
 */

// TODO — handle TypeName and SPECIAL_TYPES cases

export function generateTsAstCodeFromPgAst(ast: any): any {
    function createAstNode(functionName: string, properties: any) {
      const args = properties.map(([propKey, propValue]) => {
        return t.objectProperty(t.identifier(propKey), getValueNode(propValue));
      });
      return t.callExpression(
        t.memberExpression(t.identifier('ast'), t.identifier(functionName)),
        [t.objectExpression(args)]
      );
    }
  
    function getValueNode(value: any) {
      if (typeof value === 'object') {
        return value === null ? t.nullLiteral() : traverse(value);
      }
      switch (typeof value) {
        case 'boolean':
          return t.booleanLiteral(value);
        case 'number':
          return t.numericLiteral(value);
        case 'string':
          return t.stringLiteral(value);
        default:
          return t.stringLiteral(String(value)); // Fallback for other types
      }
    }
  
    function traverse(node: any) {
      if (Array.isArray(node)) {
        return t.arrayExpression(node.map(traverse));
      } else if (node && typeof node === 'object') {
        const entries = Object.entries(node);
        if (entries.length === 0) return t.objectExpression([]);
  
        const [key, value] = entries[0]; // Processing one key-value pair per object
        const functionName = toSpecialCamelCase(key);
        return createAstNode(functionName, Object.entries(value));
      }
  
      return getValueNode(node);
    }
  
    return traverse(ast);
  }
  