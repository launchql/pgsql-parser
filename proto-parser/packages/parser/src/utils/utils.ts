import { Enum, Field, ReflectionObject } from '@launchql/protobufjs';
import pkg from '../../package.json';
import { PgProtoParserOptions } from '../options';
import { writeFileSync } from 'fs';
import { extname, basename } from 'path';

export const getUndefinedKey = (enumName) => {
  // Split the name into parts where a lowercase letter is followed by an uppercase letter
  const parts = enumName.split(/(?<=[a-z])(?=[A-Z])/);

  const processedParts = parts.map(part => {
    // For parts that are all uppercase and longer than 1 character, only the first character should remain uppercase
    if (part === part.toUpperCase() && part.length > 1) {
      return part.charAt(0) + part.slice(1).toLowerCase();
    }
    return part;
  });

  const upperSnakeCase = processedParts.join('_').toUpperCase();

  return `${upperSnakeCase}_UNDEFINED`;
}

export const toSpecialCamelCase = (s) => {
  return s
      .replace(/_+/g, '') // Remove all underscores
      .replace(/([A-Z]+)([A-Z][a-z]|$)/g, (match, p1, p2) => 
          p1.toLowerCase() + p2 // Lowercase all but the last letter of consecutive caps
      )
      .replace(/^./, match => match.toLowerCase()); // Ensure the first character is lowercase
}

export const hasUndefinedInitialValue = (enumData: Enum) => {
  const entries = Object.entries(enumData.values);
  if (entries.length === 0) return false;

  const undefinedKey = getUndefinedKey(enumData.name);
  const firstEntry = entries[0];
  return firstEntry[0] === undefinedKey && firstEntry[1] === 0;
}

export const cloneAndNameNode = (node: ReflectionObject, name: string) => {
  const clone = JSON.parse(JSON.stringify(node));
  return {
    name,
    ...clone
  }
}

export const getFieldName = (field: Field, fallbackName: string) => {
 return field.options?.json_name ? field.options.json_name : fallbackName;
}

export const getHeader = () => {
  const version = process.env.NODE_ENV === 'test' ? 'latest' : pkg.version;
  return `/**
* This file was automatically generated by pg-proto-parser@${version}.
* DO NOT MODIFY IT BY HAND. Instead, modify the source proto file,
* and run the pg-proto-parser generate command to regenerate this file.
*/
`;
};

export const writeFileToDisk = (path: string, contents: string, options: PgProtoParserOptions) => {
  const c = (options.includeHeader && extname(path) === '.ts') ? `${getHeader()}${contents}` : contents;
  writeFileSync(path, c);
}

export const stripExtension = (filename) => {
  const extension = extname(filename);
  return basename(filename, extension);
}

