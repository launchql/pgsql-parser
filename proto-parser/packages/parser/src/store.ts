import { Service, Type, Field, Enum, Namespace, ReflectionObject } from '@launchql/protobufjs';
import { generateEnumImports, generateAstHelperMethods, generateTypeImportSpecifiers, generateEnumValueFunctions, convertEnumToTsUnionType, convertEnumToTsEnumDeclaration, generateNodeUnionType, convertTypeToTsInterface, convertTypeToWrappedTsInterface } from './ast';
import { RuntimeSchemaGenerator } from './runtime-schema';
import { generateEnum2IntJSON, generateEnum2StrJSON } from './ast/enums/enums-json';
import { jsStringify } from 'strfy-js';
import { sync as mkdirp } from 'mkdirp';
import { join } from 'path';
import { getOptionsWithDefaults, PgProtoStoreOptions } from './options';
import { cloneAndNameNode, convertAstToCode, createDefaultImport, getUndefinedKey, hasUndefinedInitialValue, stripExtension, writeFileToDisk } from './utils';
import { nestedObjCode } from './inline-helpers';
import * as t from '@babel/types';
import { NODE_TYPE } from './constants';

interface IProtoStore {
  options: PgProtoStoreOptions;
  root: ReflectionObject;
  services: Service[];
  types: Type[];
  fields: Field[];
  enums: Enum[];
  namespaces: Namespace[];

  _parse(node: ReflectionObject): void;
}
export class ProtoStore implements IProtoStore {
  options: PgProtoStoreOptions;
  root: ReflectionObject;
  services: Service[];
  types: Type[];
  fields: Field[];
  enums: Enum[];
  namespaces: Namespace[];
  private _runtimeSchema?: any[];

  constructor(
    root: ReflectionObject,
    options: PgProtoStoreOptions = {} as PgProtoStoreOptions
  ) {
    this.options = getOptionsWithDefaults(options);

    this.root = root;
    this.services = [];
    this.types = [];
    this.fields = [];
    this.enums = [];
    this.namespaces = [];

    this._parse(this.root);
  }

  _parse(node: ReflectionObject, name: string = '') {
    if (node instanceof Service) {
      this.services.push(cloneAndNameNode(node, name));
    } else if (node instanceof Type) {
      this.types.push(cloneAndNameNode(node, name));
      node.fieldsArray.forEach(field => this.fields.push(field));
    } else if (node instanceof Enum) {
      this.enums.push(cloneAndNameNode(this._processEnum(node), name));
    }

    if (node instanceof Namespace) {
      this.namespaces.push(node);
      Object.entries(node.nested || {}).forEach(([key, child]) => {
        this._parse(child, key);
      });
    }
  }

  _processEnum(enumNode: Enum): Enum {
    const undefinedKey = getUndefinedKey(enumNode.name);
    const clone = cloneAndNameNode(enumNode, enumNode.name);
    if (!this.options.enums.removeUndefinedAt0 || !hasUndefinedInitialValue(enumNode)) {
      return clone;
    }

    const newValues = {};
    let decrement = 0;

    for (const [key, value] of Object.entries(enumNode.values)) {
      if (key === undefinedKey && value === 0) {
        decrement = 1;
        continue;
      }
      newValues[key] = (value as number) - decrement;
    }

    clone.values = newValues;
    return clone;
  }

  write() {
    // Ensure the output directory exists
    mkdirp(this.options.outDir);

    this.writeEnumsJSON();
    this.writeTypes();
    this.writeWrappedTypes();
    this.writeEnums();
    this.writeUtilsEnums();
    this.writeAstHelpers();
    this.writeRuntimeSchema();
  }

  writeEnumsJSON() {
    if (this.options.enums.json.enabled) {
      const enums2int = generateEnum2IntJSON(this.enums);
      const enums2str = generateEnum2StrJSON(this.enums);

      this.writeFile(this.options.enums.json.toIntOutFile, JSON.stringify(enums2int, null, 2));
      this.writeFile(this.options.enums.json.toStrOutFile, JSON.stringify(enums2str, null, 2));
    }
  }

  allTypesExceptNode () {
    return this.types.filter(type => type.name !== NODE_TYPE);
  }

  typesToProcess () {
    return this.types
      .filter(type => type.name !== NODE_TYPE)
      .filter(type => !this.options.exclude.includes(type.name));
  }

  enumsToProcess () {
    return this.enums
      .filter(enm => !this.options.exclude.includes(enm.name));
  }

  writeTypes() {
    if (this.options.types.enabled) {
      const typesToProcess = this.typesToProcess();
      const enumsToProcess = this.enumsToProcess();
      const node = generateNodeUnionType(this.options, typesToProcess);
      const enumImports = generateEnumImports(enumsToProcess, this.options.types.enumsSource);
      const types = typesToProcess.reduce((m, type) => {
        return [...m, convertTypeToTsInterface(type, this.options)]
      }, []);
      this.writeCodeToFile(this.options.types.filename, [
        enumImports,
        node,
        ...types
      ]);
    }
  }

  writeWrappedTypes() {
    if (this.options.types.wrapped.enabled) {
      const typesToProcess = this.typesToProcess();
      const enumImports = generateEnumImports(
        this.enumsToProcess(),
        this.options.types.wrapped.enumsSource
      );
      const node = generateNodeUnionType(this.options, typesToProcess);
      const types = typesToProcess.reduce((m, type) => {
        return [...m, convertTypeToWrappedTsInterface(type, this.options, (typeName: string) => this.isWrappedType(typeName))]
      }, []);
      this.writeCodeToFile(this.options.types.wrapped.filename, [
        enumImports,
        node,
        ...types
      ]);
    }
  }

  writeEnums() {
    if (this.options.enums.enabled) {
      this.writeCodeToFile(this.options.enums.filename, 
        this.enumsToProcess().map(enm => this.options.enums.enumsAsTypeUnion ?
          convertEnumToTsUnionType(enm) :
          convertEnumToTsEnumDeclaration(enm)
        )
      );
    }
  }

  writeUtilsEnums() {
    if (this.options.utils.enums.enabled) {
      const code = convertAstToCode(generateEnumValueFunctions(this.enumsToProcess()));
      this.writeFile(this.options.utils.enums.filename, code);
    }
  }

  writeAstHelpers() {
    if (this.options.utils.astHelpers.enabled) {
      const imports = this.options.utils.astHelpers.inlineNestedObj ?
        createDefaultImport('_o', './' + stripExtension(this.options.utils.astHelpers.nestedObjFile)) :
        createDefaultImport('_o', 'nested-obj');

      const typesToProcess = this.typesToProcess();
      const code = convertAstToCode([
        imports,
        generateTypeImportSpecifiers(typesToProcess, this.options),
        generateAstHelperMethods(typesToProcess)
      ]);

      this.writeFile(this.options.utils.astHelpers.filename, code);

      if (this.options.utils.astHelpers.inlineNestedObj) {
        this.writeFile(this.options.utils.astHelpers.nestedObjFile, nestedObjCode);
      }
    }
  }

  writeRuntimeSchema() {
    if (!this.options.runtimeSchema?.enabled) {
      return;
    }

    const generator = new RuntimeSchemaGenerator(this.root as Namespace);
    const nodeSpecs = generator.generateNodeSpecs();
    const format = this.options.runtimeSchema.format || 'json';
    const filename = this.options.runtimeSchema.filename || 'runtime-schema';

    if (format === 'json') {
      const jsonContent = JSON.stringify(nodeSpecs, null, 2);
      const outFile = join(this.options.outDir, `${filename}.json`);
      writeFileToDisk(outFile, jsonContent, this.options);
    } else if (format === 'typescript') {
      const tsContent = this.generateRuntimeSchemaTypeScript(nodeSpecs);
      const outFile = join(this.options.outDir, `${filename}.ts`);
      writeFileToDisk(outFile, tsContent, this.options);
    }
  }

  getRuntimeSchema() {
    if (!this._runtimeSchema) {
      const generator = new RuntimeSchemaGenerator(this.root as Namespace);
      this._runtimeSchema = generator.generateNodeSpecs();
    }
    return this._runtimeSchema;
  }

  isWrappedType(typeName: string): boolean {
    const schema = this.getRuntimeSchema();
    const nodeSpec = schema.find(spec => spec.name === typeName);
    return nodeSpec ? nodeSpec.wrapped : false;
  }

  generateRuntimeSchemaTypeScript(nodeSpecs: any[]): string {
    const interfaceDefinitions = [
      'export interface FieldSpec {',
      '  name: string;',
      '  type: string;',
      '  isNode: boolean;',
      '  isArray: boolean;',
      '  optional: boolean;',
      '}',
      '',
      'export interface NodeSpec {',
      '  name: string;',
      '  wrapped: boolean;',
      '  fields: FieldSpec[];',
      '}',
      ''
    ];

    const exportStatement = `export const runtimeSchema: NodeSpec[] = ${jsStringify(nodeSpecs, {
      space: 2,
      camelCase: true,
      quotes: 'single'
    })};`;

    return [
      this.options.includeHeader ? this.getHeader() : '',
      ...interfaceDefinitions,
      exportStatement
    ].filter(Boolean).join('\n');
  }

  getHeader(): string {
    return `// Generated by pg-proto-parser`;
  }

  writeFile(filename: string, content: string) {
    const file = join(this.options.outDir, filename);
    writeFileToDisk(file, content, this.options);
  }

  writeCodeToFile(filename: string, nodes: t.Node[]) {
    const code = convertAstToCode(nodes);
    const filePath = join(this.options.outDir, filename);
    writeFileToDisk(filePath, code, this.options);
  }

}

