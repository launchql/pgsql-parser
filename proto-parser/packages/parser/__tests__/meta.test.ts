import ast from '../test-utils/utils/asts';
import { generateTsAstCodeFromPgAst } from '../src/utils'
import generate from '@babel/generator';

it('AST to AST to create AST â€”Â meta ðŸ¤¯', () => {
    const selectStmt = ast.selectStmt({
        targetList: [
          ast.resTarget({
            val: ast.columnRef({
              fields: [ast.aStar()]
            })
          })
        ],
        fromClause: [
          ast.rangeVar({
            relname: 'some_amazing_table',
            inh: true,
            relpersistence: 'p'
          })
        ],
        limitOption: 'LIMIT_OPTION_DEFAULT',
        op: 'SETOP_NONE'
      });

      expect(selectStmt).toMatchSnapshot();
      
      const astForAst = generateTsAstCodeFromPgAst(selectStmt);
      expect(generate(astForAst).code).toMatchSnapshot();
      expect(generate(astForAst).code).toEqual(
`ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})]
    })
  })],
  fromClause: [ast.rangeVar({
    relname: "some_amazing_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})`);

});