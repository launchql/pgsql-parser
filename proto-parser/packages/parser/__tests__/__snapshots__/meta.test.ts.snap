// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST to AST to create AST â€”Â meta ðŸ¤¯ 1`] = `
{
  "SelectStmt": {
    "fromClause": [
      {
        "RangeVar": {
          "inh": true,
          "relname": "some_amazing_table",
          "relpersistence": "p",
        },
      },
    ],
    "limitOption": "LIMIT_OPTION_DEFAULT",
    "op": "SETOP_NONE",
    "targetList": [
      {
        "ResTarget": {
          "val": {
            "ColumnRef": {
              "fields": [
                {
                  "A_Star": {},
                },
              ],
            },
          },
        },
      },
    ],
  },
}
`;

exports[`AST to AST to create AST â€”Â meta ðŸ¤¯ 2`] = `
"ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})]
    })
  })],
  fromClause: [ast.rangeVar({
    relname: "some_amazing_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;
